### DISCUSSION TITLE
Maximum size for script file

### DISCUSSION THREAD

**Discussion by Antonia-place** [Infineon/micropython#62] [General]
Hello,
What is the maximum size of code (script) file supported by MicroPython for PSoC6 in this moment? Because I have a file with aprox 1000 lines of code and the when I try to save this file into PSoC, only first aprox 350 line was saved and then I see this error in Thonny:

"Could not write next block after having written 16384 bytes to /Program.py
Make sure your device's filesystem has enough free space. (When overwriting a file, the old content may occupy space until the end of the operation.)

PROBLEM IN THONNY'S BACK-END: Exception while handling 'write_file' (OSError: [WinError Traceback (most recent call last):

  File "<stdin>", line 1, in <module>

  File "<stdin>", line 5, in __W

OSError: 28"

I know that MicroPython is in early stage of porting, but still...

Thank you,
Mihai

**Comment by NikhitaR-IFX** [comment]
Hi @Antonia-place ,

The external flash in PSoC has 512MB and the scripts you try to execute ideally is consuming this part. The above error message indicates you don't have enough space in your external flash. Two things to check here would be:
1. Is your filesystem having enough space? It could have happened that your previous applications did not clear up the space right. A tip: In order to check free space, you can use the os commands.
2. The script you are running is probably taking up a lot more space than expected. To investigate this, it matters what exactly is within your script.

If you can provide us your script, we will try to reproduce the same at our end and support better. Let us know!

Thanks and Regards,
Nikhita

**Comment by Antonia-place** [comment]
Hi,
I can not post the entire cod, because I am working on it for an Hackster contest (https://www.hackster.io/contests/connectthingswithcode).
But, I make some memory verification, in paralel, on PSoC6 and PRI Pico and I see the following data, in this pictures:
![OS memory verification_1](https://github.com/Infineon/micropython/assets/10450789/b7d88e09-83bf-494c-8523-af201c1946b8)
![The same file but from RPI Pico](https://github.com/Infineon/micropython/assets/10450789/d30d3f0a-fd8d-426b-9bd1-a671892de899)

I created the same file on RPI PIco, and I saved with success into 4MB flash; then I try to save (not run!) the same file into PSoC6 (with MicroPython v0.4.0) without success.
![No space available](https://github.com/Infineon/micropython/assets/10450789/664bf2d9-d924-4125-93db-5330c169de48)

My question is: it is an issue regarding to files with more than, let say, 500 lines on MicroPython v0.4.0 for PSoC6? Why is this difference bettween os.stat("") and os.stat("Program.pt") ( see above pictures)?

Thank you,
Mihai

**Reply by NikhitaR-IFX** [reply]
Hi,
Thank you for sharing the details.
Firstly, let me answer you on difference you observe in os.stat(). 
- os.stat("") is returning you the information about the current working directory and
- os.stat("Program.py") returns the information about this specific file meaning size of file, file permissions, modified time etc.

Talking about the issue:
From the images, I see that in Pico, the available block is 313 while in psoc6 it is 247 and I suspect that could be the reason. Nevertheless, your file size being 16384 bytes this should be enough space. Let me investigate more on this and get back to you.

**Comment by Antonia-place** [comment]
Hi,
I test again this issue, this time with a file with all lines comented and I have the same issue.
I think that is ts an issue aith the maximul file size on ported version on MIcroPython on PSOC6.
![Files size saved on PC](https://github.com/Infineon/micropython/assets/10450789/f616aa9d-e0ba-4488-bc7c-fe63b239a70c)
![Save the Test_py file with 528 comented lines](https://github.com/Infineon/micropython/assets/10450789/8d8d329e-efc3-4459-87bc-a6c4ef7791ff)
![Test_py file size after first save try](https://github.com/Infineon/micropython/assets/10450789/04571c6e-62ce-428f-9d36-889814c7e0c7)


The code file used for this test:
[Can not save large files on PSoC6.zip](https://github.com/Infineon/micropython/files/11968258/Can.not.save.large.files.on.PSoC6.zip)

Could you, please, try to reproduce this issue?

Thank you,
Mihai

**Reply by NikhitaR-IFX** [reply]
Thanks for sharing the file. I will test and let you know!

**Reply by NikhitaR-IFX** [reply]
Hi,

 I could reproduce the same issue at my end. This seems to be problem with how file allocation happens and not the size of file, as I am able to save two files with 16KB each perfectly fine. As a workaround in your actual code could you please try to break down your source code into multiple modules and import them using "import module" statement and confirm back if this works. Meanwhile we will check for a more elegant solution for this.

**Reply by Antonia-place** [reply]
Hi NikhitaR,

In gerenal it is working to split the cod in pieces, but in my case, particularry for this project, it is a nightmare, because I have to declare pins and variables at the begining of the script and then I have to use it in some "splitted" scripts...the same with sone fucntions...I have to "re-think" the whole project into a "splitted" project...This is not necessary a wrong way to approach, but until now I used a single file for whole my code.
I will try to solve this issue, but please, try to increase the block size as in RPI Pico.

Thank you for support,
Mihai

**Reply by NikhitaR-IFX** [reply]
Thanks for confirming Antonia-place and for your active contribution. We are already looking into this and you will know about the fix.

Thanks and Regards,
Nikhita

**Reply by NikhitaR-IFX** [reply]
Hi,

We figured out this is a problem with thonny IDE. I could save the file you shared on psoc6 filesystem using Arduino Lab for MPY and using mpremote tool. Please try using the latest version of [Arduino Lab for MPY](https://labs.arduino.cc/en/labs/micropython) or [mpremote tool from MPY](https://docs.micropython.org/en/latest/reference/mpremote.html).

Thanks and Regards,
Nikhita

**Reply by Antonia-place** [reply]
Hello,
Today I tested the upload of the attached test file and the result is not satisfactory: only 492 lines from the files are uploaded using Arduino Lab, the same as Thonny.
But Arduino Lab did not complain for the file size...
Moreover, sometimes, only zero lines was uploaded using Arduino LAB.

So, please verify on your side if the file is uploaded completly;
The "Test.py" contain 528 lines of code.

I attached pictures from my trys:
[Test.zip](https://github.com/Infineon/micropython/files/12039496/Test.zip)
![Only 492 lines are saved from first try](https://github.com/Infineon/micropython/assets/10450789/86f6667b-0abf-4fcd-9854-562fd992d038)
![Zero lines saved second try](https://github.com/Infineon/micropython/assets/10450789/b2f8e947-ab9b-46ec-846f-50613dccd07b)
![again 492 lines for the third try](https://github.com/Infineon/micropython/assets/10450789/226f4933-7133-485e-8ed7-00e0cd346058)

I am waiting for news,

Thank you,
Mihai

**Reply by NikhitaR-IFX** [reply]
Hi,

I saved the file you shared in the zip previously. Please see the screenshot here.

![image](https://github.com/Infineon/micropython/assets/96177627/e5b15a04-e64a-4ac8-8604-b71b15b98fcf)

Are you using the latest version of Arduino Lab for MPY? I had same issue with previous version and seems like they have upgraded their IDE. I would suggest please try with latest version of IDE or mpremote tool and confirm.

Thanks and Regards,
Nikhita

**Reply by Antonia-place** [reply]
Hi,
Previusly I used Arduinl LAB v0.5.0 .
Now, I used the latest version of Arduino LAB, 0.8.0 and indeed, the file is uploaded completly.

But I am sad that Thonny is not working, because the UI of Thonny and how all the files are handled is more intuitive than in Arduino LAB...
Maybe, in the future, Thonny will be updated and prepared for PSoC6 as is used with RPI Pico.

Thank you for support,
Mihai

**Reply by NikhitaR-IFX** [reply]
I agree completely! Thonny has been really good so far. Meanwhile we have raised issue in Thonny issues corner and hopefully we will figure out a fix soon. Thanks again for your contributions!

Thanks and Regards,
Nikhita

**Reply by Antonia-place** [reply]
Hello,
I tested again this issue using Arduinl Lab v0.8.0 and MicroPython for PSoC7 v0.5.0. 
For test, I use the attached file. 
The result is: transfer file from PC to PSoC6 remain at 99% for few seconds and then the empty files is saved into PSoC6; please see the attachments.
In this situation, I think that is not an Thonny or Arduinl Lab issue; for sure is a porting issue.
Could you, please, try to analize this?
Thank you, 
Mihai

File used for test: [D_Format_text_fix_mare.zip](https://github.com/Infineon/micropython/files/12920847/D_Format_text_fix_mare.zip)
![Remain at 99%](https://github.com/Infineon/micropython/assets/10450789/d3f98f65-8ed9-4660-9e25-960875d6c950)

![Empty file copied](https://github.com/Infineon/micropython/assets/10450789/bc69e81c-8dfa-4cf9-a4d6-fa1144f7a59d)

**Comment by jaenrig-ifx** [comment]
Solved in #68
